### PR TITLE
[Merged by Bors] - feat: add zeta_sub_one_prime

### PR DESCRIPTION
**Pull request by riccardobrasca**
We add `zeta_sub_one_prime`: in a `p^(k+1)`-th cyclotomic extension `ζ-1` is prime.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 1]
```diff
@@ -274,6 +275,68 @@ theorem subOneIntegralPowerBasis'_gen [hcycl : IsCyclotomicExtension {p} ℚ K]
   @subOneIntegralPowerBasis_gen p 1 K _ _ _ _ (by convert hcycl; rw [pow_one]) (by rwa [pow_one])
 #align is_primitive_root.sub_one_integral_power_basis'_gen IsPrimitiveRoot.subOneIntegralPowerBasis'_gen
 
+/-- `ζ - 1` is prime if `p ≠ 2` and `ζ` is a primitive `p ^ (k + 1)`-th root of unity.
+  See `zeta_sub_one_prime` for a general statement. -/
+theorem zeta_sub_one_prime_ne_two [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
+    (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) (hodd : p ≠ 2) :
+    Prime (⟨ζ - 1, Subalgebra.sub_mem _ (hζ.isIntegral (p ^ _).pos)
```
Comment (alreadydone) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:282]: Why not make a definition/abbrev for ζ as an element in the subalgebra (of algebraic integers?) and use subtraction in the subalgebra?

Comment (riccardobrasca) [reply Mathlib/NumberTheory/Cyclotomic/Rat.lean:282]: In most application `ζ` will be an element of `K` (given by `zeta`). Indeed something I can add right now is that the generator of [IsPrimitiveRoot.subOneIntegralPowerBasis](https://leanprover-community.github.io/mathlib4_docs/Mathlib/NumberTheory/Cyclotomic/Rat.html#IsPrimitiveRoot.subOneIntegralPowerBasis) is prime.

Comment (alreadydone) [reply Mathlib/NumberTheory/Cyclotomic/Rat.lean:282]: Can you make an def/abbrev `IsPrimitiveRoot.toIntegral : 𝓞 K := ⟨ζ , hζ.isIntegral (p ^ _).pos⟩` so you can write `hζ.toIntegral - 1` which should be defeq to the long expression
```
⟨ζ - 1, Subalgebra.sub_mem _ (hζ.isIntegral (p ^ _).pos) (Subalgebra.one_mem _)⟩ : 𝓞 K
```

Comment (riccardobrasca) [reply Mathlib/NumberTheory/Cyclotomic/Rat.lean]: Ah yes, using an abbreviation seems a good idea.

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 2]
```diff
@@ -181,11 +182,14 @@ noncomputable def integralPowerBasis [IsCyclotomicExtension {p ^ k} ℚ K]
   (Algebra.adjoin.powerBasis' (hζ.isIntegral (p ^ k).pos)).map hζ.adjoinEquivRingOfIntegers
 #align is_primitive_root.integral_power_basis IsPrimitiveRoot.integralPowerBasis
 
+/-- Abbreviation to see a primitive root of unity as a membmer of the ring of integers. -/
+abbrev toInteger {k : ℕ+} (hζ :IsPrimitiveRoot ζ k) : 𝓞 K := ⟨ζ, hζ.isIntegral k.pos⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:186]:
```suggestion
abbrev toInteger {k : ℕ+} (hζ : IsPrimitiveRoot ζ k) : 𝓞 K := ⟨ζ, hζ.isIntegral k.pos⟩
```

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 3]
```diff
@@ -269,11 +273,78 @@ noncomputable def subOneIntegralPowerBasis' [hcycl : IsCyclotomicExtension {p} 
...
+  rw [Nat.irreducible_iff_prime, Ideal.absNorm_span_singleton, ← Nat.prime_iff,
+    ← Int.prime_iff_natAbs_prime]
+  cases k
+  · convert Prime.neg Int.prime_two
+    apply RingHom.injective_int (algebraMap ℤ ℚ)
+    rw [← Algebra.norm_localization (Sₘ := K) ℤ (nonZeroDivisors ℤ), Subalgebra.algebraMap_eq]
+    simp only [Nat.zero_eq, PNat.pow_coe, id.map_eq_id, RingHomCompTriple.comp_eq, RingHom.coe_coe,
+      Subalgebra.coe_val, algebraMap_int_eq, map_neg, map_ofNat]
+    simpa using hζ.pow_sub_one_norm_two (cyclotomic.irreducible_rat (by simp))
+  convert Nat.prime_iff_prime_int.1 Nat.prime_two
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:319]: Can this use [`Int.prime_two`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Prime.html#Int.prime_two)?

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 4]
```diff
@@ -181,11 +182,14 @@ noncomputable def integralPowerBasis [IsCyclotomicExtension {p ^ k} ℚ K]
   (Algebra.adjoin.powerBasis' (hζ.isIntegral (p ^ k).pos)).map hζ.adjoinEquivRingOfIntegers
 #align is_primitive_root.integral_power_basis IsPrimitiveRoot.integralPowerBasis
 
+/-- Abbreviation to see a primitive root of unity as a membmer of the ring of integers. -/
```
Comment (sgouezel) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:185]:
```suggestion
/-- Abbreviation to see a primitive root of unity as a member of the ring of integers. -/
```

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 5]
```diff
@@ -269,11 +273,77 @@ noncomputable def subOneIntegralPowerBasis' [hcycl : IsCyclotomicExtension {p} 
...
     (hζ : IsPrimitiveRoot ζ p) :
-    hζ.subOneIntegralPowerBasis'.gen =
-      ⟨ζ - 1, Subalgebra.sub_mem _ (hζ.isIntegral p.pos) (Subalgebra.one_mem _)⟩ :=
+    hζ.subOneIntegralPowerBasis'.gen = hζ.toInteger - 1 :=
   @subOneIntegralPowerBasis_gen p 1 K _ _ _ _ (by convert hcycl; rw [pow_one]) (by rwa [pow_one])
 #align is_primitive_root.sub_one_integral_power_basis'_gen IsPrimitiveRoot.subOneIntegralPowerBasis'_gen
 
+/-- `ζ - 1` is prime if `p ≠ 2` and `ζ` is a primitive `p ^ (k + 1)`-th root of unity.
+  See `zeta_sub_one_prime` for a general statement. -/
+theorem zeta_sub_one_prime_ne_two [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
```
Comment (jcommelin) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:282]:
```suggestion
theorem zeta_sub_one_prime_of_ne_two [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
```

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 6]
```diff
@@ -269,11 +273,77 @@ noncomputable def subOneIntegralPowerBasis' [hcycl : IsCyclotomicExtension {p} 
...
+  convert Nat.prime_iff_prime_int.1 hp.out
+  apply RingHom.injective_int (algebraMap ℤ ℚ)
+  rw [← Algebra.norm_localization (Sₘ := K) ℤ (nonZeroDivisors ℤ), Subalgebra.algebraMap_eq]
+  simp only [PNat.pow_coe, id.map_eq_id, RingHomCompTriple.comp_eq, RingHom.coe_coe,
+    Subalgebra.coe_val, algebraMap_int_eq, map_natCast]
+  exact hζ.sub_one_norm_prime_ne_two (Polynomial.cyclotomic.irreducible_rat (PNat.pos _)) hodd
+
+/-- `ζ - 1` is prime if `ζ` is a primitive `2 ^ (k + 1)`-th root of unity.
+  See `zeta_sub_one_prime` for a general statement. -/
+theorem two_pow_zeta_sub_one_prime [IsCyclotomicExtension {(2 : ℕ+) ^ (k + 1)} ℚ K]
```
Comment (jcommelin) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:301]:
```suggestion
theorem zeta_sub_one_prime_of_two_pow [IsCyclotomicExtension {(2 : ℕ+) ^ (k + 1)} ℚ K]
```

File: Mathlib/NumberTheory/Cyclotomic/Rat.lean [hunk 7]
```diff
@@ -269,11 +273,77 @@ noncomputable def subOneIntegralPowerBasis' [hcycl : IsCyclotomicExtension {p} 
...
+theorem zeta_sub_one_prime' [h : IsCyclotomicExtension {p} ℚ K] (hζ : IsPrimitiveRoot ζ p) :
+    Prime ((hζ.toInteger - 1)) := by
+  convert zeta_sub_one_prime (k := 0) (by simpa)
+  simpa
+
+theorem subOneIntegralPowerBasis_gen_prime [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
+    (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) :
+    Prime hζ.subOneIntegralPowerBasis.gen := by simpa using hζ.zeta_sub_one_prime
+
+theorem subOneIntegralPowerBasis_gen_prime' [IsCyclotomicExtension {p} ℚ K]
```
Comment (jcommelin) [inline Mathlib/NumberTheory/Cyclotomic/Rat.lean:343]:
```suggestion
theorem subOneIntegralPowerBasis'_gen_prime [IsCyclotomicExtension {p} ℚ K]
```

**Comment by jcommelin** [comment]
bors d+

**Comment by riccardobrasca** [comment]
bors merge

**Pull request closed without merging**